components/tidb_query_expr/src/impl_regexp.rs

```diff
@@ -0,0 +1,1023 @@
+        if !is_valid_match_type(m) {
+            return Err(box_err!("invalid match type: {}", m));
+        }
+        if m == 'c' {
+            // re2 is case-sensitive by default, so we only need to delete 'i' flag
+            // to enable the case-sensitive for the regexp.
+            flag_set.remove(&'i');
+            continue;
+        }
+        flag_set.insert(m);
+    }
+
+    let mut flag = String::new();
+    for m in flag_set {
+        flag += &m.to_string();
+    }
+
+    Ok(flag)
+}
+
+fn build_regexp<C: Collator>(pattern: &[u8], match_type: &[u8]) -> Result<Regex> {
+    let match_type = get_match_type::<C>(match_type)?;
+
+    let pattern = String::from_utf8(pattern.to_vec())?;
+    let pattern_with_tp = if !match_type.is_empty() {
+        format!("(?{}){}", match_type, pattern)
+    } else {
+        pattern
+    };
+
+    Regex::new(&pattern_with_tp).map_err(|e| box_err!("invalid regex pattern: {:?}", e))
+}
+
+fn build_regexp_from_args<C: Collator>(
+    args: &[ScalarValueRef<'_>],
+    match_idx: usize,
+) -> Result<Option<Regex>> {
+    let pattern = match args[PATTERN_IDX].as_bytes() {
+        Some(b) => b,
+        None => return Ok(None),
+    };
+
+    let match_type = if args.len() > match_idx {
+        match args[match_idx].as_bytes() {
+            Some(b) => b,
+            None => return Ok(None),
+        }
+    } else {
+        b""
+    };
+
+    build_regexp::<C>(pattern, match_type).map(|reg| Some(reg))
+}
+
+fn init_regexp_data<C: Collator, const N: usize>(expr: &mut Expr) -> Result<Option<Regex>> {
+    let children = expr.mut_children();
+    if children.len() <= PATTERN_IDX {
+        return Ok(None);
+    }
+
+    let pattern = match children[PATTERN_IDX].get_tp() {
+        ExprType::Bytes | ExprType::String => children[PATTERN_IDX].get_val(),
+        _ => return Ok(None),
+    };
+
+    let match_type = if children.len() > N {
+        match children[N].get_tp() {
+            ExprType::Bytes | ExprType::String => children[N].get_val(),
+            _ => return Ok(None),
+        }
+    } else {
+        b""
+    };
+
+    build_regexp::<C>(pattern, match_type).map(|reg| Some(reg))
+}
+
+/// Currently, TiDB only supports regular expressions for utf-8 strings.
+/// See https://dev.mysql.com/doc/refman/8.0/en/regexp.html#function_regexp-like.
+#[rpn_fn(nullable, raw_varg, min_args = 2, max_args = 3, capture = [metadata], metadata_mapper = init_regexp_data::<C, LIKE_MATCH_IDX>)]
+#[inline]
+pub fn regexp_like<C: Collator>(
+    metadata: &Option<Regex>,
+    args: &[ScalarValueRef<'_>],
+) -> Result<Option<i64>> {
+    let expr = match args[0].as_bytes() {
+        Some(e) => String::from_utf8(e.to_vec())?,
+        None => return Ok(None),
+    };
+    let regex = match metadata {
+        Some(r) => r.clone(),
+        None => match build_regexp_from_args::<C>(args, LIKE_MATCH_IDX)? {
+            Some(r) => r,
+            None => return Ok(None),
+        },
+    };
```

Comment:
In which case do we go into this part?

Reply:
Let's try to avoid the unnecessary `r.clone()`. One way is change `regex` to a `Cow`, so you can have both owned or borrowed one.

Reply:
I think about it before I notice the size of `regexp` is just 16 bytes.

Reply:
![image](https://user-images.githubusercontent.com/17217495/190648059-7f952d52-fb4c-4e7d-90f0-5fab60e429b7.png)
But actually, it has a `Box<Pool>` inside. It is heavy to create that pool.

Reply:
My bad. I haven't read the clone method. 
Addressed.